Assets/Scripts/Edges/EdgeGameObjectFactory.cs

```diff
@@ -1,76 +1,66 @@
 ﻿using System.Collections.Generic;
 using System.Linq;
-using Assets.Scripts.Common;
 using Assets.Scripts.Model;
 using UnityEngine;
+using UnityEngine.UI;
 
 namespace Assets.Scripts.Edges
 {
   public class EdgeGameObjectFactory : MonoBehaviour
   {
-    private const float _additionalColliderLength = 0.1f;
-    private const float _edgeColliderRadius = 0.09f;
-    public GameObject edgePrefab;
+    const float c_edgeThicknessMultiplier = 0.1f;
+    public GameObject labelPrefab;
     public LineFactory lineFactory;
+    public MaterialCache _materialCache;
+    public void Start()
+    {
+      _materialCache = FindObjectOfType<MaterialCache>();
+    }
 
-    public void CreateGameObjectEdgesFor(IList<Edge> edges, bool labelVisibility)
+    public List<(SegmentGroup, GameObject)> CreateGameObjectEdgesFor(IList<Edge> edges, bool labelVisibility)
     {
+      if (_materialCache == null)
+        _materialCache = FindObjectOfType<MaterialCache>();
+
       var startingPoint = edges.First().nodeFrom.Position;
       var endingPoint = edges.First().nodeTo.Position;
       var edgePointsForEachEdge = lineFactory.GetLinePositionsFor(startingPoint, endingPoint, edges.Count);
+      var segmentGroups = new List<(SegmentGroup, GameObject)>();
       for (var i = 0; i < edges.Count; i++)
       {
         var edge = edges[i];
-        if (edge.gameObject != null)
-          Destroy(edge.gameObject);
-        CreateNewEdgeGameObjectFor(edge, edgePointsForEachEdge[i]);
-        edge.Text.text = edge.label;
-        edge.Text.enabled = labelVisibility;
-      }
-    }
+        var edgePoints = edgePointsForEachEdge[i];
+        
+        var segmentGameObjects = Enumerable
+          .Range(0, edgePoints.Count - 1)
+          .Select(_ => CreateCylinder(edge.Class?.Thickness ?? 1f))
+          .ToArray();
```

Comment:
I know that this is concise, but still, a regular for loop would be a lot clearer

Reply:
I disagree that a loop would be better but yeah, this is not readable. I will try to make it nicer, with a loop if necessary